### ISSUE TITLE
Any flatpak from flathub is getting SSL error

### ISSUE THREAD

**Issue by HenriqueCSJ** [flatpak/flatpak#2293]
## Linux distribution and version
Fedora 29 x86_64
## Flatpak version
Flatpak 1.0.4
## Description of the problem
Any flatpak from flathub is a nightmare to install or update. It fails 98% of the time with an error message:
```
[henrique@dhcppc3 ~]$ flatpak install flathub org.inkscape.Inkscape
Required runtime for org.inkscape.Inkscape/x86_64/stable (runtime/org.gnome.Platform/x86_64/3.28) found in remote flathub
Do you want to install it? [y/n]: y
Installing in system:
org.gnome.Platform/x86_64/3.28             flathub bc98106ca173
org.gnome.Platform.Locale/x86_64/3.28      flathub 2823e3d81b74
org.inkscape.Inkscape/x86_64/stable        flathub 87403c2e431d
  permissions: ipc, x11
  file access: host
org.inkscape.Inkscape.Locale/x86_64/stable flathub 201a189fc063
Is this ok [y/n]: y
Installing: org.gnome.Platform/x86_64/3.28 from flathub
[##                  ] Downloading files: 65/5274 755.1 kB (251.7 kB/s)
error: Failed to install org.gnome.Platform/x86_64/3.28: While pulling runtime/org.gnome.Platform/x86_64/3.28 from remote flathub: While fetching https://dl.flathub.org/repo/objects/18/f7454b9e6aba3074e0c5e7a5d16071dbc5651177eaf1fd33e452413e44dd58.filez: [56] Failure when receiving data from the peer
```
## Steps to reproduce
Try to install or update anything.

**Comment by mwleeds** [comment]
Since Flathub uses a CDN it might be helpful if you specify roughly where you are in the world

**Comment by HenriqueCSJ** [comment]
Hi! Thanks for the reply.
I'm from Brazil, Rio de Janeiro, and the error happens with a normal DSL connection of with a high-speed optic fiber in my job.

**Comment by HenriqueCSJ** [comment]
Still getting error when using flatpak:
```
[henrique@dhcppc6 ~] $ flatpak install flathub com.spotify.Client
Required runtime for com.spotify.Client/x86_64/stable (runtime/org.freedesktop.Platform/x86_64/18.08) found in remote flathub
Do you want to install it? [y/n]: y
Installing in system:
org.freedesktop.Platform/x86_64/18.08              flathub a7738ded72d8
org.freedesktop.Platform.Locale/x86_64/18.08       flathub 2e2113c42338
org.freedesktop.Platform.VAAPI.Intel/x86_64/18.08  flathub ff6f5fd65d8f
org.freedesktop.Platform.html5-codecs/x86_64/18.08 flathub dfcf5ab1cbd4
com.spotify.Client/x86_64/stable                   flathub 747830fc58b7
  permissions: ipc, network, pulseaudio, x11, dri
  file access: xdg-music:ro, xdg-pictures:ro
  dbus access: org.freedesktop.Notifications, org.gnome.SessionManager, org.gnome.SettingsDaemon
  dbus ownership: org.mpris.MediaPlayer2.spotify
  tags: proprietary
Is this ok [y/n]: y
Installing: org.freedesktop.Platform/x86_64/18.08 from flathub
[#=                  ] Downloading files: 29/12281 2,0 MB (284,0 kB/s)  
error: Failed to install org.freedesktop.Platform/x86_64/18.08: While pulling runtime/org.freedesktop.Platform/x86_64/18.08 from remote flathub: While fetching https://dl.flathub.org/repo/objects/1b/48ca8cdedb04615d5d4d06083220e32f64b879e3ba5e15e28d9e1b3521ba7e.filez: [35] SSL connect error
```

**Comment by renatobellotti** [comment]
Same here in Switzerland from Fedora 28.

**Comment by renatobellotti** [comment]
Any update on the issue?

**Comment by caioraposo** [comment]
Also from Brazil, having the same issue, no matter how many times I try.

**Comment by rochamorelli** [comment]
Same here on Silverblue 29, same place as caioraposo

**Comment by mjiricka** [comment]
Same on Arch Linux from Czech republic. I tried it many times in many days. My ISP is very mainstream. I managed to download what I wanted when I switched to other network using VPN. This way I download mayor portion of packages, then the error happened again. I turned off the VPN a was lucky to finish the download on my home connection.

**Comment by SHalintHa** [comment]
Same here. Whenever I try to install anything from flatpak (flathub)

**Comment by wswartzendruber** [comment]
Same problem. Boise, Idaho, USA.

**Comment by Pshemas** [comment]
On my end (Poland, Warsaw) I can install most of the apps from flathub, but it fails when it comes to Nvidia runtime (which basically means most of the flatpaked apps on my end stopped working):

> Installing: org.freedesktop.Platform.GL.nvidia-430-14/x86_64/1.4 from flathub
> [#=                  ] Downloading files: 1/1 569 bytes         
> error: While downloading https://us.download.nvidia.com/XFree86/Linux-x86_64/430.14/NVIDIA-Linux-x86_64-430.14-no-compat32.run: Error writing data to TLS socket: The specified session has been invalidated for some reason.

more on that here:
https://github.com/flathub/org.freedesktop.Platform.GL.nvidia/issues/23
https://github.com/flatpak/flatpak/issues/1925

I'm running Opensuse Leap 15 (Flatpak 1.0.5).

**Comment by Pshemas** [comment]
I've been able to update Nvidia by downgrading gnutls and libgnutls to older version. But I get connection / SSL errors within Flatpaked apps - even when switched back to the latest version found in repos.
So this workaround is still extremely problematic - and not available for all systems. I could do it on Leap 15 workstation, but couldn't on Leap 15.1 and Fedora 30 as there's no "old" gnutls to fall back to.

**Comment by julk3772** [comment]
yo tengo el mismo problema en Deepin

**Comment by toger5** [comment]
Any workaround for that? I have the same issue from germany.

**Comment by kurobeats** [comment]
Popped up again, this time in Australia.

**Comment by pvaret** [comment]
This looks like a duplicate of #2331 (which contains better debugging and should be the authoritative tracking bug IMO).

**Comment by tayyabsayyad** [comment]
Its still not solved from Manjaro, India, 

Error: While pulling runtime/org.kde.kdenlive.Locale/x86_64/stable from remote flathub: TLS support is not available
Error: While pulling app/org.kde.kdenlive/x86_64/stable from remote flathub: TLS support is not available
Error: While pulling app/us.zoom.Zoom/x86_64/stable from remote flathub: TLS support is not available
Updates complete.
error: There were one or more errors

**Comment by mwleeds** [comment]
@tayyabsayyad that looks like an issue with the way Flatpak was installed in your distribution. What distribution? How did you install it?

**Comment by tayyabsayyad** [comment]
It worked with updates, Thanks Mwleeds !

**Comment by shieldwed** [comment]
I know this comment is old, but regarding all the other issues, and the issue I am experiencing, this seems the one closest related.

I've been trying to install/update Slack and Spotify for months now on Arch Linux and always run into the following issue:

`error: Failed to install com.spotify.Client: While downloading https://api.snapcraft.io/api/v1/snaps/download/pOBIoZ2LrCB3rDohMxoYGnbN14EHOgD7_89.snap: While fetching https://api.snapcraft.io/api/v1/snaps/download/pOBIoZ2LrCB3rDohMxoYGnbN14EHOgD7_89.snap: [35] SSL connect error
`

Having recently tried a different distribution (TUXEDO OS, Ubuntu based), I found it working properly.

For some reason I came to the conclusion, that since flatpak is using cURL and in turn cURL is using OpenSSL by default on Arch Linux, this could be the issue. And indeed, using cURL with GnuTLS (Arch Linux package `libcurl-gnutls`) by preloading it to flatpak it is working now. This is the command to use:

`LD_PRELOAD=/usr/lib/libcurl-gnutls.so.4 flatpak install spotify`

Somehow the security software for `api.snapcraft.io` seems to block OpenSSL based requests. As this is a work around, I would like to share it for other users experiencing this issue. However I am quite uncertain where the actual bug is to be searched at.
